storage/core/src/main/java/com/hortonworks/streamline/storage/impl/jdbc/JdbcStorageManager.java

```diff
@@ -61,14 +63,17 @@ public JdbcStorageManager(QueryExecutor queryExecutor) {
     @Override
     public void add(Storable storable) throws AlreadyExistsException {
         log.debug("Adding storable [{}]", storable);
-        final Storable existing = get(storable.getStorableKey());
-
-        if(existing == null) {
-            addOrUpdate(storable);
-        } else if (!existing.equals(storable)) {
-            throw new AlreadyExistsException("Another instance with same id = " + storable.getPrimaryKey()
-                    + " exists with different value in namespace " + storable.getNameSpace()
-                    + ". Consider using addOrUpdate method if you always want to overwrite.");
+
+        try {
+            queryExecutor.insert(storable);
+        } catch (StorageException ex) {
+            if (ex.getCause() instanceof SQLIntegrityConstraintViolationException) {
+                throw new AlreadyExistsException("Another instance with same id = " + storable.getPrimaryKey()
```

Comment:
If we are OK to tolerate false positive like this, how about selecting again when SQLIntegrityConstraintViolationException is thrown?

If "select query" succeeds, we could treat previous exception as unexpected exception since the exception would mean foreign key or unique key violation.
If "select query" fails, we could treat it as AlreadyExistsException. 

There's false positive here: "insert query" fails with PK uniqueness violation, and entity is deleted from other before executing "select query". It's a tiny chance so don't hurt much, and even we can retry insertion at that time.